src/x509_extensions.rs

```diff
@@ -1,6 +1,514 @@
+    // EDIPartyName { name_assigner: Option<&'a str>, party_name: &'a str },
+    /// An uniform resource identifier. The format is not checked.
+    URI(&'a str),
+    /// An ip address, provided as encoded.
+    IPAddress(&'a [u8]),
+    RegisteredID(Oid<'a>),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct NameConstraints<'a> {
+    pub permitted_subtrees: Option<Vec<GeneralSubtree<'a>>>,
+    pub excluded_subtrees: Option<Vec<GeneralSubtree<'a>>>,
+}
+
+#[derive(Debug, PartialEq)]
+/// Represents the structure used in the name contraints extensions.
+/// The fields minimum and maximum are not supported (openssl also has no support).
+pub struct GeneralSubtree<'a> {
+    pub base: GeneralName<'a>,
+    // minimum: u32,
+    // maximum: Option<u32>,
+}
+
+pub(crate) mod parser {
+    use nom::{alt, eof, exact, do_parse, opt, verify, call, take, many1, IResult, Err};
+    use der_parser::{*, oid::Oid};
+    use der_parser::ber::BerObject;
+    use der_parser::der::*;
+    use der_parser::error::BerError;
+
+    use crate::x509_extensions::*;
+
+    pub(crate) fn parse_extension_type<'a>(mut i: &'a [u8], oid: &Oid) -> IResult<&'a [u8], Option<ExtensionType<'a>>, BerError> {
+        let ext = if oid.asn1() == oid!(raw 2.5.29.15) {
+            let (ret, ku) = parse_keyusage(i)?;
+            i = ret;
+            Some(ExtensionType::KeyUsage(ku))
+        } else if oid.asn1() == oid!(raw 2.5.29.17) {
+            let (ret, san) = parse_subjectalternativename(i)?;
+            i = ret;
+            Some(ExtensionType::SubjectAlternativeName(san))
+        } else if oid.asn1() == oid!(raw 2.5.29.19) {
+            let (ret, bc) = parse_basicconstraints(i)?;
+            i = ret;
+            Some(ExtensionType::BasicConstraints(bc))
+        } else if oid.asn1() == oid!(raw 2.5.29.30) {
+            let (ret, nm) = parse_nameconstraints(i)?;
+            i = ret;
+            Some(ExtensionType::NameConstraints(nm))
+        } else if oid.asn1() == oid!(raw 2.5.29.32) {
+            let (ret, cp) = parse_certificatepolicies(i)?;
+            i = ret;
+            Some(ExtensionType::CertificatePolicies(cp))
+        } else if oid.asn1() == oid!(raw 2.5.29.33) {
+            let (ret, pm) = parse_policymappings(i)?;
+            i = ret;
+            Some(ExtensionType::PolicyMappings(pm))
+        } else if oid.asn1() == oid!(raw 2.5.29.36) {
+            let (ret, pc) = parse_policyconstraints(i)?;
+            i = ret;
+            Some(ExtensionType::PolicyConstraints(pc))
+        } else if oid.asn1() == oid!(raw 2.5.29.37) {
+            let (ret, eku) = parse_extendedkeyusage(i)?;
+            i = ret;
+            Some(ExtensionType::ExtendedKeyUsage(eku))
+        } else if oid.asn1() == oid!(raw 2.5.29.54) {
+            let (ret, iap) = parse_inhibitanyplicy(i)?;
+            i = ret;
+            Some(ExtensionType::InhibitAnyPolicy(iap))
+        } else {
+            i = &[];
+            None
```

Comment:
Instead of an `Option<ExtensionType>`, I think adding an `Unknown( SomeType<'a> )` would be better.

- it does not require to check the option every time, only the time
- it makes this entire function easier to read and cleaner IMHO